src/sources/journald.rs

```diff
@@ -0,0 +1,281 @@
+                        );
+                        break;
+                    }
+                };
+                if self.units.len() > 0 {
+                    // Make sure the systemd unit is exactly one of the specified units
+                    if let Some(unit) = record.get("_SYSTEMD_UNIT") {
+                        if !self.units.contains(unit) {
+                            continue;
+                        }
+                    } else {
+                        continue;
+                    }
+                }
+                match channel.send(record).wait() {
+                    Ok(_) => {}
+                    Err(()) => error!(message = "Could not send journald log"),
+                }
+            }
+            // FIXME: Checkpoint here
+            match self.shutdown.recv_timeout(timeout) {
+                Ok(()) => unreachable!(), // The sender should never actually send
+                Err(RecvTimeoutError::Timeout) => {}
+                Err(RecvTimeoutError::Disconnected) => return,
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::test_util::{block_on, shutdown_on_idle};
+    use futures::stream::Stream;
+    use std::io::Error;
+    use std::iter::FromIterator;
+    use std::time::{Duration, SystemTime};
+    use stream_cancel::Tripwire;
+    use tokio::util::FutureExt;
+
+    #[derive(Default)]
+    struct FakeJournal {
+        records: Vec<Record>,
+    }
+
+    impl FakeJournal {
+        fn push(&mut self, unit: &str, message: &str) {
+            let timestamp = SystemTime::now()
+                .duration_since(SystemTime::UNIX_EPOCH)
+                .expect("Calculating time stamp failed");
+            let mut record = Record::new();
+            record.insert("MESSAGE".into(), message.into());
+            record.insert("_SYSTEMD_UNIT".into(), unit.into());
+            record.insert(
+                "_SOURCE_REALTIME_TIMESTAMP".into(),
+                format!("{}", timestamp.as_micros()),
+            );
+            self.records.push(record);
+        }
+    }
+
+    impl Iterator for FakeJournal {
+        type Item = Result<Record, Error>;
+        fn next(&mut self) -> Option<Self::Item> {
+            self.records.pop().map(|item| Ok(item))
+        }
+    }
+
+    fn fake_journal() -> FakeJournal {
+        let mut journal = FakeJournal::default();
+        journal.push("unit.service", "unit message");
+        journal.push("sysinit.target", "System Initialization");
+        journal
+    }
+
+    fn run_journal(units: &[&str]) -> Vec<Event> {
+        let (tx, rx) = futures::sync::mpsc::channel(10);
+        let (trigger, tripwire) = Tripwire::new();
+
+        let journal = fake_journal();
+        let source = journald_source(
+            journal,
+            tx,
+            HashSet::from_iter(units.into_iter().map(|&s| s.into())),
+        );
+        let mut rt = tokio::runtime::Runtime::new().unwrap();
```

Comment:
this should use `test_util::runtime()` so that we dont spawn extra threads that are not needed.

Reply:
IIRC I copied this from the file source, and a quick grep shows this pattern in almost everything else. I'll happily change the code, but you should probably set up a new issue for all the other uses.

Reply:
Yeah, I will open an issue for that, I replaced only some as they were causing issues but we should in the future use that specific runtime.